pkg/karmadactl/cmdinit/kubernetes/deploy.go

```diff
@@ -744,3 +760,229 @@ func generateServerURL(serverIP string, nodePort int32) (string, error) {
+func (i *CommandInitOption) parseEtcdConfig(etcd initConfig.Etcd) {
+	if etcd.Local != nil {
+		i.parseLocalEtcdConfig(etcd.Local)
+	} else if etcd.External != nil {
+		i.parseExternalEtcdConfig(etcd.External)
+	}
+}
+
+// parseLocalEtcdConfig parses the local Etcd settings, including image information,
+// data path, PVC size, and node selector labels.
+func (i *CommandInitOption) parseLocalEtcdConfig(localEtcd *initConfig.LocalEtcd) {
+	setIfNotEmpty(&i.EtcdImage, localEtcd.CommonSettings.Image.BuildImageString())
+	setIfNotEmpty(&i.EtcdInitImage, localEtcd.InitImage.BuildImageString())
+	setIfNotEmpty(&i.EtcdHostDataPath, localEtcd.DataPath)
+	setIfNotEmpty(&i.EtcdPersistentVolumeSize, localEtcd.PVCSize)
+
+	if len(localEtcd.NodeSelectorLabels) != 0 {
+		i.EtcdNodeSelectorLabels = mapToString(localEtcd.NodeSelectorLabels)
+	}
+
+	setIfNotEmpty(&i.EtcdStorageMode, localEtcd.StorageMode)
+	setIfNotZeroInt32(&i.EtcdReplicas, localEtcd.Replicas)
+}
+
+// parseExternalEtcdConfig parses the external Etcd configuration, including CA file,
+// client certificates, and endpoints.
+func (i *CommandInitOption) parseExternalEtcdConfig(externalEtcd *initConfig.ExternalEtcd) {
+	setIfNotEmpty(&i.ExternalEtcdCACertPath, externalEtcd.CAFile)
+	setIfNotEmpty(&i.ExternalEtcdClientCertPath, externalEtcd.CertFile)
+	setIfNotEmpty(&i.ExternalEtcdClientKeyPath, externalEtcd.KeyFile)
+
+	if len(externalEtcd.Endpoints) > 0 {
+		i.ExternalEtcdServers = strings.Join(externalEtcd.Endpoints, ",")
+	}
+	setIfNotEmpty(&i.ExternalEtcdKeyPrefix, externalEtcd.KeyPrefix)
+}
+
+// parseControlPlaneConfig parses the configuration for various control plane components,
+// including API Server, Controller Manager, Scheduler, and Webhook.
+func (i *CommandInitOption) parseControlPlaneConfig(components initConfig.KarmadaComponents) {
+	i.parseKarmadaAPIServerConfig(components.KarmadaAPIServer)
+	i.parseKarmadaControllerManagerConfig(components.KarmadaControllerManager)
+	i.parseKarmadaSchedulerConfig(components.KarmadaScheduler)
+	i.parseKarmadaWebhookConfig(components.KarmadaWebhook)
+	i.parseKarmadaAggregatedAPIServerConfig(components.KarmadaAggregatedAPIServer)
+	i.parseKubeControllerManagerConfig(components.KubeControllerManager)
+}
+
+// parseKarmadaAPIServerConfig parses the configuration for the Karmada API Server component,
+// including image and replica settings, as well as advertise address.
+func (i *CommandInitOption) parseKarmadaAPIServerConfig(apiServer *initConfig.KarmadaAPIServer) {
+	if apiServer != nil {
+		setIfNotZeroInt32(&i.KarmadaAPIServerNodePort, apiServer.Networking.Port)
+		setIfNotEmpty(&i.Namespace, apiServer.Networking.Namespace)
+		setIfNotEmpty(&i.KarmadaAPIServerImage, apiServer.CommonSettings.Image.BuildImageString())
+		setIfNotZeroInt32(&i.KarmadaAPIServerReplicas, apiServer.CommonSettings.Replicas)
+		setIfNotEmpty(&i.KarmadaAPIServerAdvertiseAddress, apiServer.AdvertiseAddress)
+	}
+}
```

Comment:
Have you seen how kubeadm does this?

Reply:
In kubeadm, this step is automatically completed by some code generation tools provided by the Kubernetes official. The tool used here is conversion-gen, and the specific code link is as follows (https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta3/zz_generated.conversion.go). The reason I did not use the code generated by this tool is that the automatically generated conversion methods are related to the structure of the structs, JSON fields, and the field names of the elements within the structs. Since the CommandInitOption structure in karmadactl differs significantly from the KarmadaInitConfig structure used for parsing the configuration file, it seems unsuitable to use the code generation tool to generate the code.